sdk/scheduler/src/main/java/com/mesosphere/sdk/debug/PlansTracker.java

```diff
@@ -0,0 +1,297 @@
+
+import com.mesosphere.sdk.http.ResponseUtils;
+import com.mesosphere.sdk.offer.Constants;
+import com.mesosphere.sdk.scheduler.plan.Phase;
+import com.mesosphere.sdk.scheduler.plan.Plan;
+import com.mesosphere.sdk.scheduler.plan.PlanCoordinator;
+import com.mesosphere.sdk.scheduler.plan.PlanManager;
+import com.mesosphere.sdk.scheduler.plan.Step;
+import com.mesosphere.sdk.state.StateStore;
+import com.mesosphere.sdk.state.StateStoreUtils;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Response;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+// SUPPRESS CHECKSTYLE MultipleStringLiteralsCheck
+
+/**
+ * PlansTracker is the backend of the PlansDebugEndpoint.
+ * It aggregates the status across all Plans and allows
+ * for filtered responses. The PlansTracker also returns
+ * the currently active plans and the current state the
+ * the scheduler is in.
+ */
+public class PlansTracker implements DebugEndpoint {
+
+  private final StateStore stateStore;
+
+  private final PlanCoordinator planCoordinator;
+
+  public PlansTracker(PlanCoordinator planCoordinator, StateStore stateStore) {
+    this.planCoordinator = planCoordinator;
+    this.stateStore = stateStore;
+  }
+
+  private JSONObject generateServiceStatus(String filterPlan,
+                                           String filterPhase,
+                                           String filterStep,
+                                           boolean allowFiltering,
+                                           boolean skipDetails)
+  {
+    JSONObject outcome = new JSONObject();
+    JSONArray plansArray = new JSONArray();
+
+    for (PlanManager planManager : planCoordinator.getPlanManagers()) {
+      Plan plan = planManager.getPlan();
+
+      //Filter down to a plan if specified.
+      if (allowFiltering && filterPlan != null && !plan.getName().equalsIgnoreCase(filterPlan)) {
+        continue;
+      }
+
+      JSONObject planObject = new JSONObject();
+      JSONArray phaseArray = new JSONArray();
+      // SUPPRESS CHECKSTYLE MultipleStringLiteralsCheck
+      planObject.put("name", plan.getName());
+      if (!skipDetails) {
+        // SUPPRESS CHECKSTYLE MultipleStringLiteralsCheck
+        planObject.put("status", plan.getStatus());
+        // SUPPRESS CHECKSTYLE MultipleStringLiteralsCheck
+        planObject.put("strategy", plan.getStrategy().getName());
+      }
+      planObject.put("phases", phaseArray);
+
+      if (!skipDetails) {
+        //Get a rollup aggregation of the steps.
+        int totalSteps = plan.getChildren().stream()
+            .flatMap(phase -> phase.getChildren().stream())
+            .collect(Collectors.toSet())
+            .size();
+        int completedSteps = plan.getChildren().stream()
+            .flatMap(phase -> phase.getChildren().stream())
+            .filter(step -> step.isComplete())
```

Comment:
this could be `.filter(Element::isComplete)` which seems easier read... but isn't a big deal